### PR TITLE
plugins.tf1: implement authentication

### LINKED ISSUE #5379: plugins.tf1: Permission insuffisante
### Checklist

- [X] This is a plugin issue and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed plugin issues](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22plugin+issue%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest stable release

### Description

Hi
Not sure exactly what means "permissions insuffisante" but sounds like something changed. Only LCI is still working

### Debug log

```text
/usr/local/bin/streamlink --loglevel debug "http://www.tf1.fr/tmc/direct" best                                                                                                         [cli][info] streamlink is running as root! Be careful!
[cli][debug] OS:         Linux-5.15.0-73-generic-x86_64-with-glibc2.29
[cli][debug] Python:     3.8.10
[cli][debug] Streamlink: 5.5.1
[cli][debug] Dependencies:
[cli][debug]  certifi: 2019.11.28
[cli][debug]  isodate: 0.6.0
[cli][debug]  lxml: 4.8.0
[cli][debug]  pycountry: 22.3.5
[cli][debug]  pycryptodome: 3.9.9
[cli][debug]  PySocks: 1.7.1
[cli][debug]  requests: 2.27.1
[cli][debug]  urllib3: 1.26.13
[cli][debug]  websocket-client: 1.3.2
[cli][debug] Arguments:
[cli][debug]  url=http://www.tf1.fr/tmc/direct
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][info] Found matching plugin tf1 for URL http://www.tf1.fr/tmc/direct
[plugins.tf1][debug] Found channel tmc (L_TMC)
[plugins.tf1][error] Permission insuffisante
error: No playable streams found on this URL: http://www.tf1.fr/tmc/direct
```

### PR DESCRIPTION
**Pull request by BellezaEmporium**
Resolves #5379

This PR is aimed to provide authentification basis for the website TF1.fr. A step forward would be to check the user subscriptions, as now certain streams might be pay-per-view, under the TF1MAX subscription plan. The answer is the same as if you're not logged in, will answer "Permission insuffisante".

With recommendations from @bastimeyer earlier on, on issue 5379, I've rewrote the plugin to make sure it suits the standards.

### CODE REVIEW

**Review by bastimeyer — CHANGES_REQUESTED** [review]
Thanks for the PR.

I'm just going to comment on the code. I currently don't have time to check the site and validate your plugin changes, including anything related to authentication and HTTP requests+responses.

There are also tons of linting issues which you'll have to fix. Please have a look at this short guide:
https://streamlink.github.io/developing.html

Also one question since I'm unaware of the account restrictions: the plugin covers multiple sites. Is authentication required for of them?

**Comment by BellezaEmporium** [comment]
Thanks for the comments. I'll get through them one by one and fix the necessary issues.

File: src/streamlink/plugins/tf1.py [hunk 1]
```diff
@@ -30,8 +31,73 @@
...
+    def _get_user_token(self):
+        # make the session request to get the correct cookies
+        if self._UID and self._UIDSignature and self._signatureTimestamp is not None:
+            session_res = self.session.http.post(
+                "https://www.tf1.fr/token/gigya/web",
+                data=json.dumps({
+                    "uid": self._UID,
+                    "signature": self._UIDSignature,
+                    "timestamp": int(self._signatureTimestamp),
+                    "consent_ids": ["1", "2", "3", "4", "10001", "10003", "10005", "10007", "10013", "10015", "10017", "10019", "10009", "10011", "13002", "13001", "10004", "10014", "10016", "10018", "10020", "10010", "10012", "10006", "10008"]
```
[the remark this answers is not in the document]
Comment (BellezaEmporium) [reply src/streamlink/plugins/tf1.py:90]: The consent_ids are linked to Gigya's API, a sort of link between the website and their services. While there's no documentation talking about it, You can find them when trying to make a connection to https://tf1.fr. The source code should give you "neededConsentIds".

Like in this example : 
image

The one I've added in this code / PR is from my own attempt at connecting using my personal account to tf1.fr.

**Comment by BellezaEmporium** [comment]
I've rewrote the plugin based on your recommendations. Last commit should reflect all the changes given.

**Comment by bastimeyer** [comment]
I force-pushed onto your PR branch because I decided to clean up the remaining issues myself and make a couple more changes.

- Fixed linting and code style issues
- Fixed missing `$account` plugin-metadata
- Fixed error handling
- Fixed validation schemas
- Fixed auth logic issues
- Renamed `--tf1-username` to `--tf1-email`
- Added token caching and `--tf1-purge-credentials`

----

Authentication is only required on `tf1.fr`. The other sites work fine, even without a French IP address:

```
$ streamlink https://tf1info.fr/direct/ best
[cli][info] Found matching plugin tf1 for URL https://tf1info.fr/direct/
[cli][info] Available streams: 234p (worst), 360p, 576p_alt, 576p, 720p (best)
[cli][info] Opening stream: 720p (hls-multi)
[cli][info] Starting player: /usr/bin/mpv
[utils.named_pipe][info] Creating pipe streamlinkpipe-153664-1-5029
[utils.named_pipe][info] Creating pipe streamlinkpipe-153664-2-3929
[cli][info] Player closed
[cli][info] Stream ended
[cli][info] Closing currently open stream...

$ streamlink https://www.lci.fr/direct best
[cli][info] Found matching plugin tf1 for URL https://www.lci.fr/direct
[cli][info] Available streams: 234p (worst), 360p, 576p_alt, 576p, 720p (best)
[cli][info] Opening stream: 720p (hls-multi)
[cli][info] Starting player: /usr/bin/mpv
[utils.named_pipe][info] Creating pipe streamlinkpipe-153955-1-8455
[utils.named_pipe][info] Creating pipe streamlinkpipe-153955-2-2202
[cli][info] Player closed
[cli][info] Stream ended
[cli][info] Closing currently open stream...
```

New token with throwaway-account:

```
$ streamlink https://www.tf1.fr/tf1/direct best --http-proxy socks5h://localhost:1920 --tf1-email ogbeyafi@email1.io --tf1-password Qwerty123456789!
[cli][info] Found matching plugin tf1 for URL https://www.tf1.fr/tf1/direct
[plugins.tf1][info] Acquiring new user-authentication token...
[cli][info] Available streams: 234p (worst), 360p, 576p_alt, 576p, 720p (best)
[cli][info] Opening stream: 720p (hls-multi)
[cli][info] Starting player: /usr/bin/mpv
[utils.named_pipe][info] Creating pipe streamlinkpipe-154141-1-4629
[utils.named_pipe][info] Creating pipe streamlinkpipe-154141-2-6526
[cli][info] Player closed
[cli][info] Stream ended
[cli][info] Closing currently open stream...
```

Cached token:

```
$ streamlink https://www.tf1.fr/tf1/direct best --http-proxy socks5h://localhost:1920
[cli][info] Found matching plugin tf1 for URL https://www.tf1.fr/tf1/direct
[cli][info] Available streams: 234p (worst), 360p, 576p_alt, 576p, 720p (best)
[cli][info] Opening stream: 720p (hls-multi)
[cli][info] Starting player: /usr/bin/mpv
[utils.named_pipe][info] Creating pipe streamlinkpipe-154271-1-2504
[utils.named_pipe][info] Creating pipe streamlinkpipe-154271-2-3907
[cli][info] Player closed
[cli][info] Stream ended
[cli][info] Closing currently open stream...

$ streamlink https://www.tf1.fr/stream/chante-69061019 best --http-proxy socks5h://localhost:1920
[cli][info] Found matching plugin tf1 for URL https://www.tf1.fr/stream/chante-69061019
[cli][info] Available streams: 234p (worst), 270p, 360p, 576p (best)
[cli][info] Opening stream: 576p (hls-multi)
[cli][info] Starting player: /usr/bin/mpv
[utils.named_pipe][info] Creating pipe streamlinkpipe-154476-1-4314
[utils.named_pipe][info] Creating pipe streamlinkpipe-154476-2-3647
[cli][info] Player closed
[cli][info] Stream ended
[cli][info] Closing currently open stream...
```

Invalid auth data:

```
$ streamlink https://www.tf1.fr/tf1/direct best --http-proxy socks5h://localhost:1920 --tf1-purge-credentials --tf1-email ogbeyafi@email1.io --tf1-password invalid
[cli][info] Found matching plugin tf1 for URL https://www.tf1.fr/tf1/direct
[plugins.tf1][info] Removing cached user-authentication token...
[plugins.tf1][info] Acquiring new user-authentication token...
error: error_code=403042 - invalid loginID or password
```

**Comment by bastimeyer** [comment]
The cache currently doesn't set an expiration time/date, but it's not really important...

Going to merge this now.

**Comment by BellezaEmporium** [comment]
Perfect, thanks for going further in this PR. Interesting, thought they would enforce it everywhere. I consider myself corrected.

**Comment by octavioj** [comment]
This is wonderful. Which version of streamline will have this modification, if I may ask?

**Comment by bastimeyer** [comment]
`6.7.4+1.g7adb5b12`
https://github.com/streamlink/streamlink/commits/master/

**Comment by octavioj** [comment]
Thank you.

**Comment by campones** [comment]
I tested few hours ago the beta release from yesterday, works perfectly

**Comment by octavioj** [comment]
It works wonderfully with liveproxy too. Excellent.

**Comment by campones** [comment]
@octavioj sorry for the off topic but what's the point to use this liveproxy exactly?

**Comment by BellezaEmporium** [comment]
That's not the correct place to talk about this I believe.

**Comment by campones** [comment]
probably not. anyway, good job guys,  has been quite a long way but result is outstanding ;)

**Pull request merged**